File: day_03/day3.jl
#=
day3:
- Julia version: 
- Author: mike.downey
- Date: 2022-12-04
=#

# file = open("day_03/test.txt")
file = open("day_03/data.txt")
data = readlines(file)

# Part 1
println("Part 1")

# Pass a Char (not a String) and get the priority
function priority(letter)
    asc = 1 + Int(letter) - Int('a')
    if asc > 0
        return asc
    else
        return 27 + Int(letter) - Int('A')
    end
end

function halves(contents)
    return [contents[begin:div(length(contents),2)], contents[1+div(length(contents),2):end]]
end

# Returns a set of Char
s2c = s -> s[1]
function string2set(contents)
    letters = split(contents,"")
    chars = s2c.(letters) # dot notation applies function to elements of an array
    return Set(chars)
end

function get_in_common(row)
    half = halves(row)
    return intersect(string2set(half[1]), string2set(half[2]))
end

function get_priority_for_common(row)
    return sum(priority.(get_in_common(row)))
end

println(sum(get_priority_for_common.(data)))

# Part 2
# Group into blocks of 3 and find their badge types, sum those.
println("Part 2")

function find_badge_and_priority(elves)
    allhave = intersect(elves[1], elves[2], elves[3])
    return priority(collect(allhave)[1])
end

priorities = [find_badge_and_priority(data[i:i+2]) for i in range(1, length(data), step=3)]
println(sum(priorities))

File: day_03/day3.py
from aoc import aoc


def priority(letter):
    asc = 1 + ord(letter) - ord("a")
    return asc if asc > 0 else 27 + ord(letter) - ord("A")


def halves(contents):
    return contents[:len(contents) // 2], contents[len(contents) // 2:]


def string2set(letters):
    s = set()
    for l in letters:
        s.add(l)
    return s


def sum_priorities(items):
    return sum([priority(item) for item in items])


def get_in_common(row):
    first, second = halves(row)
    return string2set(first).intersection(string2set(second))


def get_priority_for_common(row):
    return sum_priorities(get_in_common(row))


def part1(data):
    return sum([get_priority_for_common(datum) for datum in data])


def part2(data):
    priorities = 0
    for i in range(0, len(data), 3):
        elves = data[i:i+3]
        allhave = set(elves[0]) & set(elves[1]) & set(elves[2])
        priorities += priority(list(allhave)[0])
    return priorities


def run(file):
    print("Part 1")
    data = aoc.loadlines(file)
    print(part1(data))
    print("Part 2")
    print(part2(data))


if __name__ == "__main__":
    print("Test Data")
    run("test.txt")
    print("Actual Data")
    run("data.txt")
